### PR TITLE
[MBUILDCACHE-33] Allow to set remote cache credentials from env vars

### PR DESCRIPTION
**Pull request by AlexanderAshitkin**
Introduces set of environment variables to provide remote cache settings from environment variables. 
The feature is to support cases when developers do not have access to settings xml and lack access to build infrastructure (managed build containers)

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MNG) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[MNG-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `MNG-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.
 - [x] You have run the [Core IT][core-its] successfully.

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [x] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

[core-its]: https://maven.apache.org/core-its/core-it-suite/

### CODE REVIEW

**Review by michael-o — COMMENTED** [review]
I have expressed my objections in the JIRA issue and why I consider this wrong.

**Comment by AlexanderAshitkin** [comment]
> I have expressed my objections in the JIRA issue and why I consider this wrong.

Discussed in Jira. There are strong reasons why using credentials from settings.xml in CI environment creates either security risks either unnecessary complicates maintenance. I'm happy to implement better solution as long as solution which addresses these absolutely basic concerns offered.

File: src/main/java/org/apache/maven/buildcache/RemoteCacheRepositoryImpl.java [hunk 1]
```diff
@@ -81,11 +90,58 @@ public RemoteCacheRepositoryImpl(
...
+                    proxy = new Proxy( uri.getScheme(), uri.getHost(), uri.getPort() );
+                }
+            }
+            else
+            {
+                proxy = session.getProxySelector().getProxy( repo );
+            }
+
+            Authentication auth;
+            if ( env.containsKey( MAVEN_BUILD_CACHE_USER ) )
```
Comment (gnodet) [inline src/main/java/org/apache/maven/buildcache/RemoteCacheRepositoryImpl.java:132]: `&& env.containsKey( MAVEN_BUILD_CACHE_PASSWORD )` ?  Or cleanly fail if it's missing maybe.

File: src/main/java/org/apache/maven/buildcache/RemoteCacheRepositoryImpl.java [hunk 2]
```diff
@@ -81,11 +90,58 @@ public RemoteCacheRepositoryImpl(
...
+            }
+
+            // if proxy is set by environment, use it
+            Proxy proxy;
+            if ( env.containsKey( MAVEN_BUILD_CACHE_PROXY_URL ) )
+            {
+                String proxyUrl = env.get( MAVEN_BUILD_CACHE_PROXY_URL );
+                LOGGER.debug( "Remote build cache proxy url overridden by environment to {}", proxyUrl );
+                URI uri = URI.create( proxyUrl );
+                if ( env.containsKey( MAVEN_BUILD_CACHE_PROXY_USER ) )
```
Comment (gnodet) [inline src/main/java/org/apache/maven/buildcache/RemoteCacheRepositoryImpl.java:112]: `&& env.containsKey( MAVEN_BUILD_CACHE_PASSWORD )` ?  Or cleanly fail if it's missing maybe.

**Review by gnodet — CHANGES_REQUESTED** [review]

**Comment by michael-o** [comment]
I still consider this wrong and credentials must come from `settings.xml`. There you can use env var interpolation.

**Comment by rmannibucau** [comment]
+1 to make settings.xml optional, it creates issues or complicated maintenance for nothing - leading to set it fully in env vars soletimes - but ultimatly it should be configurable from the env in setting reader (without materializing the file) to benefit maven as a whole and not require specific config each time IMHO.
Can we hope to work toward that?

**Comment by gnodet** [comment]
@AlexanderAshitkin have you tried using environment variable interpolation from `settings.xml` ?
Also it would be nice to add a bit of documentation about available solutions (settings.xml with or without interpolation / env vars).

**Comment by AlexanderAshitkin** [comment]
> @AlexanderAshitkin have you tried using environment variable interpolation from `settings.xml` ?
> Also it would be nice to add a bit of documentation about available solutions (settings.xml with or without interpolation / env vars).

No, I haven't. Also I can't modify settings in xml in my current project, so it will not work for me in any case.

**Comment by AlexanderAshitkin** [comment]
@gnodet @michael-o I feel like there is no consensus for this change yet, I'm declining this pr so far. Need to agree on the design first.

**Comment by jira-importer** [comment]
Resolve #292

**Comment by jira-importer** [comment]
Resolve #292

**Pull request closed without merging**